java/ql/lib/semmle/code/java/regex/regex.qll

```diff
@@ -0,0 +1,943 @@
+import java
+private import RegexFlowConfigs
+
+/**
+ * A string literal that is used as a regular exprssion.
+ */
+abstract class RegexString extends Expr {
+  RegexString() { this instanceof StringLiteral }
+
+  /** Holds if a character set starts between `start` and `end`. */
+  private predicate char_set_start0(int start, int end) {
+    this.nonEscapedCharAt(start) = "[" and
```

Comment:
Can `[` itself occur in a character class? `[[]` for example?

Reply:
No; in `[[ ]` the `[ ]` would be considred as a nested class, making the outer class unfinished and thus a parse error.